doc/decisions/DR023-Versioning-traits-and-specifications.md

```diff
@@ -0,0 +1,702 @@
+
+
+# DR023 Versioning traits and specifications
+
+- **Status:** To be decided
+- **Impact:** High
+- **Driver:** @elliotcmorris
+- **Approver:** @feltech
+- **Contributors:** @elliotcmorris @feltech @reinecke @antirotor
+- **Outcome:**  TBC
+
+### Preface
+This document refers to the versioning of trait and specification
+libraries in general. However, as `openassetio-mediacreation` is the
+only one in play at the moment, and since reading "trait and
+specifications library" over and over would be tiresome, we will just
+use the term `mediacreation` throughout this document.
+
+# Problem Statement
+In order to achieve OpenAssetIOs interoperability goal, hosts and
+managers must agree on the structure of data exchanged between them.
+
+The `openassetio-mediacreation` defines the agreed standard for interop
+objects (called `traits` and `specifications`) across the VFX industry.
+Hosts and Managers must agree on the structure of these objects. [These
+objects](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation/blob/main/traits.yml)
+are currently in a first-pass state, and it's understood that the only
+way to come to a solid set of useful shared concepts is via use and
+iteration.
+
+OpenAssetIO wants to be thin, and does not want to be "the problem". To
+this end, we need a change and versioning process for the traits and
+specifications schema which allows maximum compatibility, especially in
+the face of the unpredictable version adoption in the VFX industry.
+
+[`openassetio-mediacreation`](https://github.com/OpenAssetIO/OpenAssetIO-MediaCreation)
+is a generative library, in that it produces trait and specification
+view classes representing the objects specified in the source `.yml` via
+the `openassetio-traitgen` tool, currently in both the c++ and python
+languages. As such, the challenge of versioning is not only descriptive,
+but also technical.
+
+# Aims :
+- Allow new mediacreation versions to be adopted by managers without
+  breaking compatibility with hosts running previous versions.
+- Deliver a solid but minimal versioning foundation, that leaves open
+  opportunities for implementation conveniences at a later date.
+- Allow infinite backwards compatibility for manager implementations.
+- Low cognitive load, the OpenAssetIO development ecosystem cannot
+  support a complicated/technically intricate solution.
+- Minimize burden on manager implementors, whilst keeping in mind the
```

Comment:
I suspect this was about our assumption that the manager would be responsible for most of the work to adapt to whatever the host advertises as its supported schema version; but we want to make that adaptation as painless as possible. For hosts it would be trivial. However, this assumption may be upended by https://github.com/OpenAssetIO/OpenAssetIO/pull/1254/files#r1482525401

Reply:
Maybe this comes down to a question of what the fundamental interaction model is. Historically, its always been that Hosts can store/recall data in/from a Manager, Managers may opt-in to layer on first-class understanding of the data. 

The rational here was that in a future (largely unrealistic) world, all data would have originated from one host or another. So the manager acts more like a database, where the ideal manager allows the host to store/recall assets that it doesn't even understand. Eg DCC -> DCC could exchange specific data without the manager needing to understand it per-se. This was a critical part of the abstraction. Not all managers may support this, but it is possible, and should be aimed for unless there is a compelling reason not to.

In practice this is probably less likely to be the case unless a manager embraces the API fully (though that has happened with USD...), the fact that the API was designed to facilitate that was a conscious choice.

  https://openassetio.github.io/OpenAssetIO/index.html#manager_implementation_concept

Management policy and others are a concession that many managers may have compromised support, but at least you can determine this accurately.

If we switch to the "host determines a version and the managers have to obey", then it pretty much turns this upside down, to a model where the manager is required to have first-class understanding of all data in any request. This would preclude host A publishing one version of a proprietary asset and host B resolving it having requested another version. If this is a vendor specific asset type, the manager may have no way to perform this migration, but the hosts - as they consume/produce that data with first class understanding - may well be able to. 

I think its the same reason that we chose not to enforce 'required' fields in the middleware, as the host is the only one who knows if a missing trait property is actually an error or not. The manager or middleware never can.

A world (probably not entirely dissimilar to OTIO?) were hosts in the ecosystem can exchange assets not-fully understood by the manger between themselves seems incredibly valuable, and seems to be readily facilitated where the data model is coherent and self-describing allows the responsibility of dealing with differing versions to lie with whoever wishes to layer on concrete understanding of that data not the broker in the middle. 

Auto-magic support could be layerd on if desired, but isn't a pre-requisite as long as it is possible to pass this through the manager as a 'hub' in the so-called hub/spoke model.

Keeping data interpretation de-coupled from its storage and retrieval feels really important if we want to make a long-term flexible interchange mechanism.

The existing API methods all traffic trait IDs, which allows any cleverness in any part host or manager to then be implemented if desired without any structural changes. eg. `resolve`  is requesting a specific version, so a host could resolve n trait versions and see which come back from a 'simple' manager. A complex manager could opt-in to doing version conversion from the requested id to the stored one it if wanted /could. I guess here per-trait versioning is beneficial as it means you dont need as much out-of-band knowledge as discussed else where.  This sounds familiart with my understanding of OTIO too (though I may well be wrong) so may also mean its not wheel re-invention?

Reply:
An alternate view of OTIO is that it has per-schema versioning, but their schemas are much smaller.

I do agree the idea of a pass-through manager with no first-class understanding of the traits is something we need to keep in mind with any solution. Also related: https://github.com/OpenAssetIO/OpenAssetIO/pull/1254#discussion_r1496051054

Reply:
> An alternate view of OTIO is that it has per-schema versioning, but their schemas are much smaller.

Yeah - though I guess I was meaning more the part of "the version is in the data" rather than "the version is ephemeral runtime metadata".